# En_CellReports_2024

This page describes codes used to analyze snRNA-seq data in En ... Ikegami, Cell Reports 2024 paper.\
All raw data reported in this paper are available at:
[GEO GSE241590](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE241590)

snRNA-seq data, including scRNA-seq matrix, are in this sub-series:
[GEO GSE241587](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE241587)

# Single nuclear RNA-seq

## 1) Getting snRNA-seq count matrix

### Get scripts from here:
[https://github.com/JunyueC/sci-RNA-seq3_pipeline/tree/master/script_folder](https://github.com/JunyueC/sci-RNA-seq3_pipeline/tree/master/script_folder)

I modified some scripts to run on my files in my environments. These are `*_KI.sh` or `*_KI.py` files. These are provided in this page. 


### Rename fastq files
```sh
#############################
# Run rename_fastq_KI.sh
################################

# Usage: rename_fastq_KI.sh [a COPY!! of fastq_dir] [sample_id]

$ ~/script_folder/rename_fastq_KI.sh ~/KICHMC008/fastq_copy/ ~/KICHMC008/fastq_copy/KICHMC008_sample_ID.txt 

```

### Attach UMI
```sh
##############################
# Make pickle2 files
#################################

# Put these files in script_folder
# 3lvl_mRNA_Lig_plate_1_bc.txt
# 3lvl_mRNA_RT_plate_1_bc.txt

$ module load python/2.7.5\npython ~/script_folder/generate_RT_lig_pickle_KI.py

###############################
# Run UMI_attach_KI.sh
##################################


# UMI_attach_KI.sh [fastq] [sample_id] [outdir] [cpu] [Lig] [RT] [script dir]

$ ~/script_folder/UMI_attach_KI.sh ~/KICHMC008/fastq_copy/ ~/KICHMC008/fastq_copy/KICHMC008_sample_ID.txt ~/KICHMC008/ 20 ~/script_folder/3lvl_mRNA_Lig_plate_1_bc.pickle2 ~/script_folder/3lvl_mRNA_RT_plate_1_bc.pickle2 ~/script_folder/"

```


### Trimming and Alignment

```sh
##################################
# Make genome index
#####################################

$ module load STAR/2.7.9
$ STAR --runMode genomeGenerate --runThreadN 8 --genomeDir ~/genome_bfa/mm39/star/without_annotation --genomeFastaFiles ~/genome_bfa/mm39/mm39.fa

#######################################
# Run trim_and_align_KI.sh
#########################################

# trim_and_align_KI.sh [umi fastq dir] [sample_id] [outdir] [star index] [cpu] [script dir]

$ ~/script_folder/trim_and_align_KI.sh ~/KICHMC008/UMI_attach ~/KICHMC008/fastq_copy/KICHMC008_sample_ID.txt ~/KICHMC008/trim_and_align/ ~/genome_bfa/mm39/star/without_annotation/ 20 ~/script_folder/

```


### Filter reads by UMI, barcodes, coordinates

```sh
###############################################
# Making combined barcode file
##################################################

$ ~/script_folder/make_combined_bc_KI.sh ~/script_folder/3lvl_mRNA_Lig_plate_1_bc.txt ~/script_folder/3lvl_mRNA_RT_plate_1_bc.txt > ~/script_folder/3lvl_mRNA_combined_bc.txt


###########################################
# Run filter_sam_KI.sh
#####################################

# Usage: filter_sam_KI.sh [STAR outdir] [sample_id] [UMI cutoff] [cpu] [barcodes] [script] [outdir]

$ ~/script_folder/filter_sam_KI.sh ~/KICHMC008/trim_and_align/STAR_alignment/ ~/KICHMC008/fastq_copy/KICHMC008_sample_ID.txt 200 20 ~/script_folder/3lvl_mRNA_combined_bc.txt ~/ ~/KICHMC008/filter_sam/sam_splitted/"

```


### Gene counting

```sh
################################
# Run gene_count_KI.sh
################################

# Usage: gene_count_KI.sh [sam_splitted] [barcode_sample] [gtf] [cpu] [script] [outdir]
# Note barcode_samples.txt is generated by filter_sam_KI.sh

$ ~/script_folder/gene_count_KI.sh ~/KICHMC008/filter_sam/sam_splitted/ ~/KICHMC008/filter_sam/barcode_samples.txt ~/genome_bfa/mm39/gene_annotation/gencode.vM27.basic.annotation.gtf 50 ~/script_folder/ ~/KICHMC008/gene_count"

```


### Make R objects

```sh
#######################################
# Install packages
###############################################

$ module load R/4.2.1

$ install.packages("Matrix");
$ install.packages("tidyverse");
$ install.packages("data.table");

###########################################
# Run gene_count_processing_sciRNAseq_KI.R
###########################################

$ module load R/3.2.0\nRscript --vanilla ~/sci-RNA-seq3_pipeline-master/gene_count_processing_sciRNAseq_KI.R \"~/KICHMC008/gene_count/\" \"~/KICHMC008/data_frame/\""

```

```r
load("~/ikegami/mouse/sciRNAseq/KICHMC008/data_frame/sci_summary.RData")

#                       Type      Size       Rows Columns
# gene_count       dgCMatrix 759925728 2791643652      NA
# df_gene         data.frame  13732416      55359       3
# df_cell         data.frame   5245512      50428       3


```

## 2) Processing snRNA-seq count matrix

### Adding cell information
```r
######################################
# Adding cell information
########################################

# Read plates and sample info
library(readxl)

p7_plate <- read_excel("PCR_P7_plate1.xlsx", sheet="PCR_P7_plate1", col_names=c("p7_well","index","sequence"), skip=1) %>%
	separate(p7_well, sep=1, into=c("p7_row","p7_col"), remove=F) %>%
	separate(index, sep="index_10nt_", into=c("blank","i7_bc")) %>%
	mutate(sequence=str_remove(sequence, "^CAAGCAGAAGACGGCATACGAGAT")) %>%
	mutate(i7_bc_seq=str_to_upper(str_remove(sequence, "GTCTCGTGGGCTCGG$"))) %>%
	select(-blank, -sequence) %>%
	left_join(sample_sheet, by=c("i7_bc" = "i7")) %>%
	select(-Pool_Name, -(blank:code))

lig_plate <- read_excel("3lvl_mRNA_Lig_plate_1.xlsx", sheet="3lvl_mRNA_Lig_plate_1", col_names=c("lig_well","index","sequence"), skip=1) %>%
	separate(lig_well, sep=1, into=c("lig_row","lig_col"), remove=F) %>%
	separate(index, sep="sc_ligation_", into=c("blank","lig_bc")) %>%
	mutate(lig_bc_seq=str_remove(sequence, "^.*TACGACGCTCTTCCGATCT")) %>%
	select(-blank, -sequence)

RT_plate <- read_excel("3lvl_mRNA_RT_plate_1.xlsx", sheet="3lvl_mRNA_RT_plate_1", col_names=c("rt_well","index","sequence"), skip=1) %>%
	separate(rt_well, sep=1, into=c("rt_row","rt_col"), remove=F) %>% 
	separate(index, sep="sc_ligation_RT_", into=c("blank","rt_bc")) %>%
	mutate(sequence=str_remove(sequence, "^/5Phos/CAGAGCNNNNNNNN")) %>%
	mutate(rt_bc_seq=str_remove(sequence, "TTTTTTTTTTTTTTTTTTTTTTTTTTTTTT$")) %>%
	select(-blank, -sequence)

my_sample <- read_excel("sample_info2.xlsx") %>%
	mutate(rt_col=as.character(rt_col))

# Update df_cell

df_cell <- df_cell %>%
	separate(sample, sep=-10, into=c("lig_bc_seq","rt_bc_seq"), remove=F) %>%
	separate(lig_bc_seq, sep="\\.", into=c("sample_id","lig_bc_seq"), remove=T) %>%
	left_join(RT_plate,  by="rt_bc_seq") %>%
	left_join(lig_plate,  by="lig_bc_seq") %>%
	left_join(p7_plate, by=c("sample_id"="sample")) %>%
	left_join(my_sample, by="rt_col")
	 
```


```r
########################################
# Making singleCellExperiment object
#######################################

BiocManager::install("scDblFinder")
BiocManager::install("SingleCellExperiment")
BiocManager::install("scater")
BiocManager::install("scran")

library(SingleCellExperiment)

load("~/ikegami/mouse/sciRNAseq/KICHMC008/data_frame/sci_summary_03-03-2023.RData")

# 1) my_sce

my_rowRanges <- split(makeGRangesFromDataFrame(df_gene %>% column_to_rownames("gene_id"),
	seqnames.field="chr",
	start.field="start",
	end.field="stop",
	strand.field="strand",
	starts.in.df.are.0based=F,
	keep.extra.columns=TRUE,
	seqinfo=Seqinfo(genome="mm39")
))


my_sce <- SingleCellExperiment(
	assays=list(counts= gene_count),
	colData=DataFrame(df_cell),
	metadata=list(study="KICHMC008"),
	rowRanges=my_rowRanges,
)

> my_sce
class: SingleCellExperiment 
dim: 55359 50428 
metadata(1): study
assays(1): counts
rownames(55359): ENSMUSG00000102693.2 ENSMUSG00000064842.3 ... ENSMUSG00000064371.1 ENSMUSG00000064372.1
rowData names(0):
colnames(50428): KI990s01.TGATGCGATGGCAGACGCC KI990s01.ACAACCTATTAAGACCGTTG ... KI990s95.AGATCGGATTTACCGAGGC
  KI990s95.GATACGTCTGGTACTGCCT
colData names(31): sample_id lig_bc_seq ... genotype2 genot_tam
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):


```


```r
###############################
# Attach basic QC using scuttle
#############################

location <- rowRanges(my_sce)
is.mito <- any(seqnames(location)=="chrM")
my_sce <- addPerCellQCMetrics(my_sce, subsets=list(Mito=is.mito))

saveRDS(my_sce, file="~/KICHMC008/R_object/KICHMC008_my_sce_03-07-2023.rds")

```


### Preprocessing 1

```r
#################################
# Define preproSce1 function
#################################

preproSce1 <- function(sce, percell_umi, percell_gene, unmatch, mito_perc, HVG_num, PCA_num, db_group, db_rate){

	library(SingleCellExperiment)
	library(Matrix)
	library(scater);
	library(scran);
	library(scDblFinder);
	library(BiocParallel);
	
	# 1) Filter by umi, gene, unmatch, and mito
	message("1) Filter by umi, gene, unmatch, and mito");
	
	sce <- sce[,(
		sce$UMI_count > percell_umi & 
		sce$detected > percell_gene &
		sce$unmatched_rate < unmatch &
		sce$subsets_Mito_percent < mito_perc
	)]
	
	# 2) Normalization
	message("2) Normalization");
	
	sce <- logNormCounts(sce)
	
	# 3) Doublet detection
	message("3) Doublet detection");
	
	bp <- MulticoreParam(4, RNGseed=19)
	set.seed(19)
	sce <- scDblFinder(
		sce, 
		samples= db_group, 
		dbr=db_rate, 
		clusters=TRUE,
		BPPARAM=bp
	) 
	
	# 4) Feature selection
	message("4) Feature selection");
	
	HVG_name <- paste0("HVG", HVG_num);
	my_var <- modelGeneVar(sce, assay.type = "logcounts");  
	rowSubset(sce, HVG_name) <- getTopHVGs(my_var, var.field="bio", n=HVG_num);
	
	# 5) PCA
	message("4) PCA");
		
	PCA_name <- paste0("PCA", PCA_num);
	
	set.seed(19)
	sce <- fixedPCA(
		sce, 
		subset.row=rowSubset(sce, HVG_name), 
		rank=PCA_num, 
		preserve.shape=TRUE, 
		assay.type="logcounts", 
		name=PCA_name
	)

	# 6) UMAP
	message("6) UMAP");

	set.seed(19)
	sce <- runUMAP(
		sce, 
		dimred=PCA_name, 
		subset_row=rowSubset(sce, HVG_name)	)

	return(sce);
}

```


```r
########################
# Run preproSce1
##########################

my_sce <- readRDS(file="~/ikegami/mouse/sciRNAseq/KICHMC008/R_object/KICHMC008_my_sce_03-07-2023.rds")

my_sce <- preproSce1(
 	sce=my_sce,
 	percell_umi=200,
	percell_gene=100,
	unmatch=0.4,
	mito_perc=10,
	HVG_num=2000,
	PCA_num=50,
	db_group="mouse_id",
	db_rate=0.1
)
```

### Preprocessing 2

```r
##############################
# Define preproSce2 function
############################

preproSce2 <- function(sce, mito_perc, db_score, HVG_num, PCA_num){

	library(scater);
	library(scran);
	
	# 1) Filter by mito (again) and double score
	message("1) Filter by umi, gene, unmatch, and mito");
	
	sce <- sce[,(
		sce$subsets_Mito_percent < mito_perc &
		sce$scDblFinder.score < db_score
	)]
	
	# 2) Normalization
	message("2) Normalization");
	
	sce <- logNormCounts(sce)
		
	# 4) Feature selection
	message("4) Feature selection");
	
	HVG_name <- paste0("HVG", HVG_num);
	my_var <- modelGeneVar(sce, assay.type = "logcounts");  
	rowSubset(sce, HVG_name) <- getTopHVGs(my_var, var.field="bio", n=HVG_num);
	
	# 5) PCA
	message("4) PCA");
		
	PCA_name <- paste0("PCA", PCA_num);
	
	set.seed(19)
	sce <- fixedPCA(
		sce, 
		subset.row=rowSubset(sce, HVG_name), 
		rank=PCA_num, 
		preserve.shape=TRUE, 
		assay.type="logcounts", 
		name=PCA_name
	)
	return(sce);
}

```


```r
#######################
# UMAP, clustering, & annotation 
###########################

library(bluster)

my_sub <- my_sce[, my_sce$day_post_tam == "14"];


# 1) PreproSce2
my_sub <- preproSce2(my_sub, mito_perc=2, db_score=0.25, HVG_num=2000, PCA_num=50)

# 2) UMAP
set.seed(19)
my_sub <- runUMAP(my_sub, dimred="PCA50", n_neighbors=20, min_dist=1, n_threads=4)


plotUMAP(my_sub, point_size=0.1) +  theme_my5(aspect_ratio=1)

# 3) Clustering
set.seed(19)

myclust2 <- clusterCells(my_sub, use.dimred="PCA50",
                         BLUSPARAM=TwoStepParam(
                           first=KmeansParam(centers=2000, iter.max=100),
                           second=NNGraphParam(k=8)
                         )
);


# 4) Evaluation by Plotting
plotUMAP(my_sub, colour_by=I(myclust2), point_size=0.1) +
    theme_my5(aspect_ratio=1, legend="top", base_size=12) +
    labs(title="Cluster, my_sub")

# 5) Name data objects

sce14111_2w <- my_sub;

## Remove previous annotation
sce14111_2w$K8 <- NULL; sce14111_2w$K12 <- NULL; sce14111_2w$K8_annot <- NULL; sce14111_2w$K12_annot <- NULL;

sce14111_2w$K12 <- myclust2;

# 6) Add annotation

k12_annotation <- data.frame(
    K12=as.character(c(1:12)),
    K12_annot=factor(
        c("Fib2","Mac2","CEC","Mac1","CM2","EC","Fib1","CM1","Tcell","Neuron","LEC","PC"),
        levels=c("CM1","CM2","Fib1","Fib2","Mac1","Mac2","Tcell","EC","CEC","LEC","PC","Neuron")
    )
) %>%
    arrange(K12_annot)

> k12_annotation
   K12 K12_annot
1    8       CM1
2    5       CM2
3    7      Fib1
4    1      Fib2
5    4      Mac1
6    2      Mac2
7    9     Tcell
8    6        EC
9    3       CEC
10  11       LEC
11  12        PC
12  10    Neuron

sce14111_2w$K12_annot <- colData(sce14111_2w) %>% adf %>%
    left_join(k12_annotation, by="K12") %>%
    pull(K12_annot)

K12br <- data.frame(
	K12_annot=c("CM1","CM2","Fib1","Fib2","Mac1","Mac2","Tcell","EC","CEC","LEC","PC","Neuron"),
	K12br_annot=factor(c("CM","CM","Fib","Fib","Imm","Imm","Imm","EC","CEC","LEC","PC","Neuron"), 
		levels=c("CM","Fib","Imm","EC","CEC","LEC","PC","Neuron")
	)
)

> K12br
   K12_annot K12br_annot
1        CM1          CM
2        CM2          CM
3       Fib1         Fib
4       Fib2         Fib
5       Mac1         Imm
6       Mac2         Imm
7      Tcell         Imm
8         EC          EC
9        CEC         CEC
10       LEC         LEC
11        PC          PC
12    Neuron      Neuron

sce14111_2w$K12br_annot <- colData(sce14111_2w) %>% adf %>%
    left_join(K12br, by="K12_annot") %>%
    pull(K12br_annot)

```

## 3) Pseudobulk DESeq2

### Prepare data and metadata

```r
#############################
# function: make_counts_ls
############################

 
make_counts_ls <- function(sce, cluster_col, sample_col) {

    cluster_names <- colData(sce)[, cluster_col] %>% levels;
    sample_names <- colData(sce)[,sample_col] %>% unique %>% as.character;
    groups <- colData(sce)[, c(cluster_col, sample_col)]
    
    library(Matrix);
    # matrix.utils has been put in computating_resources folder. and Just run the Matrix.utils.R function
	 source(file="~/ikegami/computing_resources/executables/R/Matrix.utils/R/Matrix.utils.R")
    
    aggr_counts <- aggregate.Matrix(t(counts(sce)), groupings = groups, fun = "sum");
    aggr_counts <- t(aggr_counts);
     
    ## Initiate empty list
    counts_ls <- list()
    
    for (i in 1:length(cluster_names)) {
        ## Extract indexes of columns in the global matrix that match a given cluster
        column_idx <- grep(paste0("^", cluster_names[i],"_"), colnames(aggr_counts))
        
        ## Store corresponding sub-matrix as one element of a list
        counts_ls[[i]] <- aggr_counts[, column_idx]
        names(counts_ls)[i] <- cluster_names[i]
    }
    return(counts_ls);
}

########################
# Run make_counts_ls
############################

my_counts_ls <- make_counts_ls(sce= sce14111_2w, cluster_col="K12br_annot", sample_col="mouse_id");


#############################
# function: make_metadata_ls
############################

make_metadata_ls <- function(sce, cluster_col, sample_col, counts_ls, metadata) {
	    
	t <- table(colData(sce)[,sample_col], colData(sce)[,cluster_col])
	    
	# Creating metadata list
	    
	## Initiate empty list
	metadata_ls <- list()
	    
	for (i in 1:length(counts_ls)) {
	
		message(paste("Processing", i));
		  
		## Initiate a data frame for cluster i with one row per sample (matching column names in the counts matrix)
		df <- data.frame(cluster_sample_id = colnames(counts_ls[[i]])) %>%
		    separate(cluster_sample_id, into=c("cluster_id", "sample_id"), remove=F)
		    
		## Retrieve cell count information for this cluster from global cell count table
		idx <- which(colnames(t) == unique(df$cluster_id))
		cell_counts <- t[, idx]
		    
		## Remove samples with zero cell contributing to the cluster
		cell_counts <- cell_counts[cell_counts > 0]
		    
		## Match order of cell_counts and sample_ids
		sample_order <- match(df$sample_id, names(cell_counts))
		cell_counts <- cell_counts[sample_order]
		    
		## Append cell_counts to data frame
		df$cell_count <- cell_counts
		    
		## Join data frame (capturing metadata specific to cluster) to generic metadata
		df <- left_join(df, metadata, by = c("sample_id" = sample_col))
		                     
		## Update rownames of metadata to match colnames of count matrix, as needed later for DE
		rownames(df) <- df$cluster_sample_id
		    
		## Store complete metadata for cluster i in list
		metadata_ls[[i]] <- df
		names(metadata_ls)[i] <- unique(df$cluster_id)
	}
	return(metadata_ls);

}


#############################
# Run: make_metadata_ls
############################

my_metadata <- colData(sce14111_2w) %>% 
    as.data.frame() %>% 
    dplyr::select(mouse_id, genot_tam, series, sex, day_post_tam, genotype2) %>%
    group_by(mouse_id) %>%
    distinct %>%
    mutate(rownames=mouse_id) %>%
    mutate(genotype3=factor(case_when(
        genotype2=="Lmna++" ~ "LmnaPP",
        genotype2=="LmnaFF" ~ "LmnaFF",
    ))) %>%
    column_to_rownames(var="rownames")


my_metadata_ls <- make_metadata_ls(sce= sce14111_2w, cluster_col="K12br_annot", sample_col="mouse_id", counts_ls=my_counts_ls, metadata= my_metadata);

```

### Run DESeq2

```r
#############################################
# Function get_dds_resultsAvsB  to do DESeq for each cluster
#############################################

get_dds_resultsAvsB <- function(clustx, A, B, counts_ls, metadata_ls, group_id, series) {
    
    library(pheatmap)
    library(DESeq2)
      
    print(clustx) # useful for debugging
      
    # Extract counts matrix and metadata for cluster x
    idx <- which(names(counts_ls) == clustx)
    cluster_counts <- counts_ls[[idx]]
    cluster_metadata <- metadata_ls[[idx]]
      
    # Print error message if sample names do not match
    if ( all(colnames(cluster_counts) != rownames(cluster_metadata)) ) {
        print("ERROR: sample names in counts matrix columns and metadata rows do not match!")
    }
      
    dds <- DESeqDataSetFromMatrix(cluster_counts, 
                            colData = cluster_metadata, 
                            design = as.formula(paste("~", group_id)))
      
    # Transform counts for data visualization
    rld <- rlog(dds, blind = TRUE)
      
    # Generate QC plots
      
    ## Plot and save PCA plot
    DESeq2::plotPCA(rld, intgroup = group_id) +
    theme_my5(aspect_ratio=1) +
    labs(title=paste("Cluster", clustx, series))
    
    ggsave(paste0("plot/cluster_", clustx, "_specific_PCAplot.png"))
      
    ## Extract rlog matrix from the object and compute pairwise correlation values
    rld_mat <- assay(rld);
    rld_cor <- cor(rld_mat);
      
    ## Plot and save heatmap
    pheatmap(
        rld_cor, annotation = cluster_metadata[, c(group_id), drop = FALSE],
        filename=paste0("plot/cluster_", clustx, "_specific_heatmap.png")
    )
      
    # Run DESeq2 differential expression analysis
    dds <- DESeq(dds)
      
    ## Plot dispersion estimates
    png(paste0("plot/cluster_", clustx, "_dispersion_plot.png"),
    height = 5, width = 6, units = "in", res = 300)
    plotDispEsts(dds)
    dev.off()
      
    ## Output and shrink results of Wald test for contrast A vs B
    contrast <- paste(c(group_id, A, "vs", B), collapse = "_")
    # resultsNames(dds)
      
    res <- results(dds, name = contrast, alpha = 0.05)
    res <- lfcShrink(dds, coef = contrast, res = res)
    
    message("Done lfcshrink.")
      
    ## Turn the results object into a tibble for use with tidyverse functions
    res_tbl <- res %>%
    data.frame() %>%
    rownames_to_column(var = "gene_id")

    message("Done making result table.")
          
    # Generate results visualization plots
      
    ## Extract normalized counts from dds object
    normalized_counts <- counts(dds, normalized = TRUE) %>%
        data.frame() %>%
        rownames_to_column(var = "gene_id")  

    message("Done making normalized count table.")
      
    ## Combine res_tbl with normalized counts
    final_table <- res_tbl %>% left_join(normalized_counts, by="gene_id")
    message("Done making final table.")
    
    saveRDS(final_table, file=paste0(series, "_cluster", clustx, "_DESeq_res.rds"))
    
    message("Done saving RDS.")
}

#######################################
# Run function get_dds_resultsAvsB
#########################################

setwd("~/ikegami/mouse/sciRNAseq/KICHMC008/sce14111_2w/DESeq/K12br")
cluster_names <- levels(colData(sce14111_2w)$K12br_annot)

for(i in c(1:length(cluster_names))) {
    get_dds_resultsAvsB(
        clustx=cluster_names[i],
        A = "LmnaFF", 
        B = "LmnaPP", 
        counts_ls=my_counts_ls,
        metadata_ls=my_metadata_ls,
        group_id = "genotype3",
        series="tam14d"
    )
}


```

### Process DESeq data

```r
###############################
# Process DESeq data
###############################

setwd("~/ikegami/mouse/sciRNAseq/KICHMC008/sce14111_2w/DESeq/K12br")
cluster_names <- levels(colData(sce14111_2w)$K12br_annot)

my_metadat <- colData(sce14111_2w) %>% 
    as_tibble() %>% 
    dplyr::select(mouse_id, genot_tam, series, sex, day_post_tam, genotype2) %>%
    unique %>% remove_rownames %>%
    mutate(mouse_id=as.character(mouse_id)) %>%
    mutate(Lmna=case_when(
        genotype2=="Lmna++" ~ "PP",
        genotype2=="LmnaFF" ~ "FF",
    ))

t14_res <- NULL;
for(i in c(1:length(cluster_names))) {
    
    #cluster=paste0("c",i);
    
    my_res <- readRDS(file=paste0("./tam14d_cluster", cluster_names[i], "_DESeq_res.rds"));
    
    my_res <- my_res %>%
        pivot_longer(cols=c(-gene_id, -baseMean, -log2FoldChange, -lfcSE, -pvalue, -padj), 
        	names_to="mouse_id", values_to="norm_count"
        ) %>%
        separate(mouse_id, into=c("header","mouse_id"), sep="_") %>%
        rename(log2FoldChange="LFC", pvalue="pv") %>%
        left_join(my_metadat, by="mouse_id") %>%
        group_by(gene_id, LFC, lfcSE, pv, padj, Lmna) %>%
        summarize(ncountMean=mean(norm_count)) %>%
        ungroup %>%
        pivot_wider(names_from=Lmna, values_from=ncountMean) %>%
        mutate(K12br_annot=cluster_names[i])
    
    t14_res <- rbind(t14_res, my_res)
}


sce14111_K12br_deseq <- t14_res %>%
    mutate(lfc0padj5_bin3=factor(case_when(
        is.na(LFC) | is.na(padj) ~ "no",
        LFC > 0 & padj < 0.05 ~ "up",
        LFC < 0 & padj < 0.05 ~ "dn",
        .default = "no"
    ), levels=c("up","dn","no"))) %>%
    mutate(lfc1padj5_bin3=factor(case_when(
        is.na(LFC) | is.na(padj) ~ "no",
        LFC > 1 & padj < 0.05 ~ "up",
        LFC < -1 & padj < 0.05 ~ "dn",
        .default = "no"
    ), levels=c("up","dn","no")))


# save in RDS
saveRDS(sce14111_K12br_deseq, file="~/ikegami/mouse/sciRNAseq/KICHMC008/sce14111_2w/DESeq/K12br/sce14111_K12br_deseq_07-07-2023.rds")

```




## 4) Cellchat K12

### Run CellChat

```r
####################################################
# Cell chat on K12 using truncated mean with 10% cutoff
####################################################

# 1) Replacing rownames with gene_name 
my_sce <- sce14111_2w;
rowData(my_sce)$gene_id <- rownames(my_sce);
rowData(my_sce)$gene_name <- rowRanges(my_sce) %>% as.data.frame %>% pull(gene_name);
rownames(my_sce) <- rowData(my_sce)$gene_name;

# 2) Subset by samples

my_coldat <- get_mycoldat(my_sce);
my_coldat %>% mutate(genot_tam=factor(genot_tam)) %>% dplyr::select(genot_tam) %>% summary
       genot_tam   
 Lmna++.tam14:6299  
 LmnaFF.tam14:7812

sce_tam14pp <- my_sce[, (my_coldat %>% dplyr::filter(genot_tam=="Lmna++.tam14") %>% pull(cell_id)) ]
sce_tam14ff <- my_sce[, (my_coldat %>% dplyr::filter(genot_tam=="LmnaFF.tam14") %>% pull(cell_id)) ]


preproc_cellchat2 <- function(sce, group, species, type="triMean", trim=0.1) {
	
	if(species == "mouse") {
		my_database= get_cellchatDB(species="mouse");
		my_ppi= PPI.mouse;
	}
	
	my_cc <- createCellChat(object = sce, group.by = group);
		message("Done creating cellChat object");
	my_cc@DB <- my_database;
		message("Done attaching database");
	my_cc <- subsetData(my_cc);
		message("Done initiazing cellChat object");	my_cc <- identifyOverExpressedGenes(my_cc);
		message("Done identifyOverExpressedGenes");
	my_cc <- identifyOverExpressedInteractions(my_cc);
		message("Done identifyOverExpressedInteractions");
	my_cc <- projectData(my_cc, my_ppi);
		message("Done projectData on PPI");
	my_cc <- computeCommunProb(my_cc, type=type, trim=trim);
		message("Done computeCommunProb");
	my_cc <- filterCommunication(my_cc, min.cells = 10);
		message("Done filterCommunication with min cells 10");
	my_cc <- computeCommunProbPathway(my_cc);
		message("Done computeCommunProbPathway");
	my_cc <- aggregateNet(my_cc);
		message("Done aggregateNet");
	
	return(my_cc);
}


library(CellChat);

cc_k12_tam14pp <- preproc_cellchat2(
    sce= sce_tam14pp, 
    group= "K12_annot",
    species="mouse",
    type="truncatedMean",
    trim=0.1
)

cc_k12_tam14ff <- preproc_cellchat2(
    sce= sce_tam14ff, 
    group= "K12_annot",
    species="mouse",
    type="truncatedMean",
    trim=0.1
)

save(cc_k12_tam14pp, cc_k12_tam14ff, file="~/ikegami/mouse/sciRNAseq/KICHMC008/R_object/sce14111_2w_cellchat_K12_truncMean0.1_obj_07-13-2023.RData")

```


```r
# All K12 samples

all.k12_cclist <- list(
    t14pp = cc_k12_tam14pp,
    t14ff = cc_k12_tam14ff
)

my_merge_k12_cc <- mergeCellChat(all.k12_cclist, add.names = names(all.k12_cclist));

intdat_k12 <- subsetCommunication(my_merge_k12_cc, slot.name = "net", sources.use = c(1:12), targets.use = c(1:12), thresh=0.05);

my_cells <- levels(my_merge_k12_cc@idents$t14pp);

my_pairs <- data.frame(
        source=factor(rep(my_cells, each=12), levels=my_cells),
        target=factor(rep(my_cells, times=12), levels=my_cells) 
    ) %>%
    mutate(source.target = factor(paste(source, target, sep="_"))) %>%
    mutate(source.target=fct_inorder(source.target)) 


intdat_k12_long <- bind_rows(intdat_k12, .id="sample") %>%
    mutate(
        sample = fct_inorder(sample),
        source = factor(source, levels=my_cells),
        target = factor(target, levels=my_cells),
        source.target = factor(paste(source, target, sep="_")),
        interaction_name = factor(interaction_name),
        pathway_name = factor(pathway_name),
        annotation = factor(annotation)
    ) %>%
    arrange(source, target) %>%
    mutate(source.target=fct_inorder(source.target)) %>%
    dplyr::select(sample, source.target, source, target, interaction_name, prob, pathway_name, annotation) %>%
    arrange(source.target)

```


```r
###############################################################
# K12 annotation, but analysis only on "G7" (CMs, Fibs, and Imms)
#############################################################

intdat_G7 <- subsetCommunication(my_merge_k12_cc, slot.name = "net", sources.use = c(1:7), targets.use = c(1:7), thresh=0.05);

my_cells <- levels(my_merge_k12_cc@idents$t14pp)[1:7];

my_pairs <- data.frame(
        source=factor(rep(my_cells, each=7), levels=my_cells),
        target=factor(rep(my_cells, times=7), levels=my_cells) 
    ) %>%
    mutate(source.target = factor(paste(source, target, sep="_"))) %>%
    mutate(source.target=fct_inorder(source.target)) 

intdat_G7_long <- bind_rows(intdat_G7, .id="sample") %>%
    mutate(
        sample = fct_inorder(sample),
        source = factor(source, levels=my_cells),
        target = factor(target, levels=my_cells),
        source.target = factor(paste(source, target, sep="_")),
        interaction_name = factor(interaction_name),
        pathway_name = factor(pathway_name),
        annotation = factor(annotation)
    ) %>%
    arrange(source, target) %>%
    mutate(source.target=fct_inorder(source.target)) %>%
    dplyr::select(sample, source.target, source, target, interaction_name, prob, pval, pathway_name, annotation) %>%
    arrange(source.target)

```

### Plot Cellchat interaction data

```r
#########################
# Plot G7
##########################

intdat_G7_long %>%
	mutate(source=factor(source, levels=c("CM1","CM2","Fib1","Fib2","Mac1","Mac2","Tcell"))) %>%
	mutate(target=factor(target, levels=c("CM1","CM2","Fib1","Fib2","Mac1","Mac2","Tcell"))) %>%
	dplyr::count(annotation, sample, source, target, .drop=F) %>%
	pivot_wider(names_from=sample, values_from=n) %>%
	mutate(t14_dif=t14ff-t14pp) %>%
	ggplot(
        aes(x=source, y=fct_rev(target), fill= mylimit(t14_dif,-10,25))
    ) +
    geom_tile(color="white") +
    scale_fill_distiller(palette = "Spectral") +
    scale_x_discrete(position = "top") +
    facet_wrap(~annotation, strip.position="bottom") +
    theme_my5(aspect_ratio=1, base_size=11, legend="right", xangle=90, xhjust=0) +
    labs(title="K12 G7 Difference of interaction count between FF and WT within annotation")
    
    
```

### Plot signaling mediators

```r
#####################################################
# # Plot pathway in select interactions -- Take2
####################################################

plotdat <- intdat_G7_long %>%
 	dplyr::filter(annotation != "Cell-Cell Contact") %>%
	mutate(
		annot=case_when(
			annotation=="ECM-Receptor" ~ "ECM",
			annotation=="Secreted Signaling" ~ "Sec"
		),
		source.target.annot=paste(source.target, annot, sep="_")
	) %>%
	dplyr::count(sample, source.target.annot, pathway_name) %>%
	group_by(sample, source.target.annot) %>%
	mutate(count_rank=row_number(-n)) %>%
	ungroup %>%
	mutate(
		path2=ifelse(count_rank <= 3, as.character(pathway_name), "else"),
		count_rank2=ifelse(count_rank <= 3, count_rank, 4),
	) %>%
	group_by(sample, source.target.annot, count_rank2, path2) %>%
	summarize(count=sum(n)) %>%
	ungroup %>%
	arrange(sample, source.target.annot, count_rank2, path2, count)
	
	
my_int <- c("CM2_Fib2_ECM", "Fib2_CM2_ECM", "Fib2_Fib2_ECM", "Fib2_Tcell_ECM", "Fib2_CM2_Sec", "Fib2_Fib2_Sec", "Fib2_Mac2_Sec", "Mac2_Mac2_Sec");
		
for(i in c(1:length(my_int))) {
	pdat <- plotdat %>%
		dplyr::filter(source.target.annot==my_int[i]) %>%
		mutate(path2=fct_inorder(path2)) %>%
		ggplot(
			aes(x=sample, y=count, fill=path2)
		) +
		geom_col(position="stack", color="black", linewidth=0.5) +
		scale_fill_tableau("Superfishel Stone") +
		theme_my5(aspect_ratio=1.8, legend="right", legend.width = 0.4) +
		labs(title=my_int[i]);
	
	assign(paste0("p", i), pdat)
} 


```
